add prometheus integration in recommendation-app (#19)
streamshub/flink-sql-examples: prometheus-install/README.md

# Integrate Prometheus into Flink cluster

> After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by these steps:

Reviewer: ```suggestion
After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by following these steps:
```

1. Update the prometheus configuration with the ip address of flink pods to scrape from:

   **Linux:**
   ```

>    sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml

Reviewer: ```suggestion
   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
```

I think it's likely people would run this from the top level directory

>    sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-config.yaml

Reviewer: ```suggestion
   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
```

   ```
   **MacOS**
   ```

>    sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml

Reviewer: ```suggestion
   sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
```

>    sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-config.yaml

Reviewer: ```suggestion
   sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
```

   ```
   Note: Here we assume there's only 1 job manager and 1 task manager. If you deployed more than that, please update the `prometheus-config.yaml` file.   

2. Install prometheus, configuration, and service:
   ```

>    kubectl apply -f . -n flink

Reviewer: ```suggestion
   kubectl apply -f prometheus-install -n flink
```

> # Integrate Prometheus into Flink cluster in non-openshift cluster

Reviewer: ```suggestion
# Integrate Prometheus into Flink cluster
```
I don't think we need the "in non-openshift cluster" part here since these are the instructions most people would follow

> After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by following steps:

Reviewer 2: ```suggestion
After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics from job manager and task manager by following these steps:
```

> 1. Update the prometheus configuration with the ip address of flink pods to scrape from:

Reviewer 2: ```suggestion
1. Update the prometheus configuration with the IP address of Flink pods to scrape from:
```

   **Linux:**
   ```
   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
   ```
   **MacOS**
   ```
   sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
   sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
   ```
   Note: Here we assume there's only 1 job manager and 1 task manager. If you deployed more than that, please update the `prometheus-config.yaml` file.   

2. Install prometheus, configuration, and service:
   ```

>    kubectl apply -f prometheus-install -n flink

Reviewer 2: would this command apply everything under prometheus-install directory? if yes, I think it might be confusing because this directory has flink-monitor.yaml etc. Maybe we should have a sub directory for non openshift yamls only.

Author: Oops, you're right! Let me update it.

   ```

3. Expose the prometheus UI with port-forward rule:
   ```
   kubectl port-forward svc/prometheus-service -n flink 9090
   ```

> 4. Now you can monitor the metrics in job manager or task manager via the prometheus UI is accessible at localhost:9090.

Reviewer 2: ```suggestion
4. Now you can monitor the metrics in job manager or task manager via the Prometheus UI is accessible at localhost:9090.
```

Reviewer 2: Prometheus and Flink sometimes starts with uppercase and sometimes with lowercase.

![img.png](job_metric.png)
![img.png](task_metric.png)

> # Integrate Prometheus into Flink cluster in openshift cluster

Reviewer: ```suggestion
# Integrate Prometheus into Flink cluster deployed on OpenShift
```

> Since Openshift already has built-in Prometheus installed and configured, we can integrate with it by deploying `PodMonitor` CR for flink cluster:

Reviewer: ```suggestion
Since OpenShift already has a built-in Prometheus installed and configured, we can integrate with it by deploying a `PodMonitor` CR for the flink cluster:
```

1. Install the pre-configured `PodMonitor` CR:
   ```
   oc apply -f prometheus-install/flink-monitor.yaml -n flink
   ```

>    Note: This CR is using `recommendation-app` as an example. Please update it if it's for other application.

Reviewer: ```suggestion
   Note: This CR is configured to select the `FlinkDeployment` created as part of the `recommendation-app` example. Please update the `selector.matchLabels` field in `flink-monitor.yaml` if you are running a different example.

> 2. It takes around 5 minutes to wait for prometheus operator to update the config for prometheus server. After that, you can query the metrics in the Openshift UI as described [here](https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-metrics.html#querying-metrics-for-all-projects-as-an-administrator_managing-metrics).

Reviewer: ```suggestion
2. It takes around 5 minutes to wait for prometheus operator to update the config for prometheus server. After that, you can query the metrics in the OpenShift UI as described [here](https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-metrics.html#querying-metrics-for-all-projects-as-an-administrator_managing-metrics).
```
